qubes/mgmt.py

```diff
@@ -144,37 +133,96 @@ def vm_list(self, untrusted_payload):
+        # explicit list to be sure that it matches protocol spec
+        if isinstance(property_def, qubes.vm.VMProperty):
+            property_type = 'vm'
+        elif property_def.type is int:
+            property_type = 'int'
+        elif property_def.type is bool:
+            property_type = 'bool'
+        elif self.arg == 'label':
+            property_type = 'label'
+        else:
+            property_type = 'str'
+
         try:
             value = getattr(self.dest, self.arg)
         except AttributeError:
-            return 'default=True '
+            return 'default=True type={} '.format(property_type)
         else:
-            return 'default={} {}'.format(
+            return 'default={} type={} {}'.format(
                 str(self.dest.property_is_default(self.arg)),
-                self.repr(value))
+                property_type,
+                str(value) if value is not None else '')
 
     @asyncio.coroutine
-    def vm_property_help(self, untrusted_payload):
+    def vm_property_set(self, untrusted_payload):
+        assert self.arg in self.dest.property_list()
+
+        property_def = self.dest.property_get_def(self.arg)
+        # do not treat type='str' as sufficient validation
+        if isinstance(property_def, qubes.vm.VMProperty):
+            try:
+                untrusted_vmname = untrusted_payload.decode('ascii')
+            except UnicodeDecodeError:
+                raise qubes.exc.QubesValueError
+            qubes.vm.qubesvm.validate_name(self.dest, property_def,
+                untrusted_vmname)
+            try:
+                newvalue = self.app.domains[untrusted_vmname]
+            except KeyError:
+                raise qubes.exc.QubesPropertyValueError(
+                    self.dest, property_def, untrusted_vmname)
+        elif property_def.type is not None and property_def.type is not str:
+            if property_def.type is bool:
+                try:
+                    untrusted_value = untrusted_payload.decode('ascii')
+                except UnicodeDecodeError:
+                    raise qubes.exc.QubesValueError
+                newvalue = qubes.property.bool(None, None, untrusted_value)
+            else:
+                try:
+                    newvalue = property_def.type(untrusted_payload)
+                except ValueError:
+                    raise qubes.exc.QubesValueError
+        else:
+            # other types (including explicitly defined 'str') coerce to str
+            # with limited characters allowed
+            try:
+                untrusted_value = untrusted_payload.decode('ascii')
+            except UnicodeDecodeError:
+                raise qubes.exc.QubesValueError
+            allowed_set = string.printable
+            if not all(x in allowed_set for x in untrusted_value):
+                raise qubes.exc.QubesValueError(
+                    'Invalid characters in property value')
+            newvalue = untrusted_value
```

Comment:
Not sure about that - I'd like to have all the untrusted-input validation code in one place (file), for easier auditing. If we spread that across multiple files (imagine searching for all qubes.property subclasses), that would be much harder to audit.

Maybe such validator should be _in addition_ to preliminary validation here? But on the other hand, we already have setters which are serving mostly this purpose.